docs/ADR-X-Better-call-stacks-of-io-exceptions.md

```diff
@@ -0,0 +1,79 @@
+# Status
+
+📜 Proposed 2025-03-04
+
+# Context
+
+This ADR enriches exceptions thrown from `IO` functions with call stacks up to the function usage site.
+
+## The problem
+This ADR bases on [[ADR-8-Use-RIO-in-cardano‐cli]].
+
+The goal of this ADR is to provide better call stacks in `IO` exceptions.
+The problem with exceptions thrown from IO monad is that they do not carry call stack from where they were thrown.
+This is not very convenient when for example doing multiple file access or network operations in a row.
+You won't know which one has thrown an exception: you will have only a vague exception message without many details, for example:
+```
+Network.Socket.recvBuf: resource vanished (Connection reset by peer)
+```
+
+## Solution
+
+The currently used type alias adds call stacks to the functions:
+```haskell
+type CIO e a = HasCallStack => CIO e a
+```
+We can make sure that `IO` exceptions thrown in `IO` actions are captured and rethrown with a helper `runIO` function, meant to be used instead of `liftIO`:
+
+```haskell
+import UnliftIO.Exceptions (catchAny, throwIO)
+
+runIO :: IO a -> CIO a
+runIO m = withFrozenCallStack $ catchAny (liftIO m) (throwIO . mkE)
+ where
+  mkE :: SomeException -> IoeWrapper
+  mkE e = withFrozenCallStack $ IoeWrapper e
```

Comment:
Why not do that as part of an implementation of the instance of `MonadIO` for `CIO`. That would have the advantage that we could use functions that already implement `MonadIO` without a change

Reply:
I've tried that. However there are two problems:
1. You'd have to create a new monad to write a custom MonadIO instance. Then you still have to use a type alias afterwards, because you can't just make `HasCallStack` work on a datatype this way: https://stackoverflow.com/questions/79482064/how-to-add-hascallstack-to-a-custom-monad-in-haskell
2. `liftIO` does not have `HasCallStack` in its signature, which means that it will lose the call stack to the place where it was called. I've tried that too, `InstanceSigs` don't help here either.

Reply:
Ah, I see, pity. Then it makes complete sense